src/utils/env.js

```diff
@@ -1,27 +1,34 @@
+// dotenv loading similar to create-react-app / react-scripts or dotenv-load
+
 const path = require('path')
 const dotenv = require('dotenv')
-const { statAsync, readFileAsync } = require('../lib/fs')
+const filterObject = require('filter-obj')
+const { isFileAsync, readFileAsync } = require('../lib/fs')
 
 async function getEnvSettings(projectDir) {
-  const envDevelopmentFile = path.resolve(projectDir, '.env.development')
-  const envFile = path.resolve(projectDir, '.env')
-
-  const settings = {}
-
-  try {
-    if ((await statAsync(envFile)).isFile()) settings.file = envFile
-  } catch (err) {
-    // nothing
-  }
-  try {
-    if ((await statAsync(envDevelopmentFile)).isFile()) settings.file = envDevelopmentFile
-  } catch (err) {
-    // nothing
-  }
-
-  if (settings.file) settings.vars = dotenv.parse(await readFileAsync(settings.file)) || {}
+  const dotenvFiles = ['.env.development', '.env']
+  const results = await Promise.all(
+    dotenvFiles.map(async file => {
+      const filepath = path.resolve(projectDir, file)
+      const isFile = await isFileAsync(filepath)
+      if (!isFile) {
+        return
+      }
+      const content = await readFileAsync(filepath)
+      const parsed = dotenv.parse(content)
+      // only keep envs not configured in process.env
+      const env = filterObject(parsed, key => !Object.prototype.hasOwnProperty.call(process.env, key))
+      return { file, env }
+    })
+  )
 
-  return settings
+  const settings = results.filter(Boolean).reduce(
+    ({ files, vars }, { file, env }) => {
+      return { files: [...files, file], vars: { ...env, ...vars } }
+    },
+    { files: [], vars: {} }
+  )
+  return { ...settings, vars: Object.entries(settings.vars) }
```

Comment:
[sand] I am wondering whether `Array.map()` would be clearer than `Array.reduce()`? Something like:

```js
const allResults = results.filter(Boolean)
const files = allResults.map(({ file }) => file)
const vars = Object.entries(Object.assign({}, ...allResults.map(({ env }) => env)))
return { files, vars }
```

I have not tried that code, it might be wrong :)
Feel free to dismiss if you think `Array.reduce()` is simpler :+1:

Reply:
Maybe, I was trying to save that additional iteration 😄
I think you would need to reverse the order of results when mapping to `vars` to make it work as excepted, as files are ordered by precedence.

Reply:
This is less efficient, but it might be clearer. Although that is not clear either... so please feel free to dismiss :)

Reply:
Done in https://github.com/netlify/cli/pull/1305/commits/6e2f29cd7b7ed2fe7e69643b00542ac99fbd78f6#diff-eac6f78170256c598e70ce4b121d69caL25